### PR TITLE
refactor: cleanup glob implementation, add tests

### PR DESCRIPTION
**Pull request by jbedard**
Prefactor to https://github.com/aspect-build/bazel-lib/issues/419

To add more test coverage and cleanup to keep the actual fix minimal.

---

### Type of change
- Refactor (a code change that neither fixes a bug or adds a new feature)

### Test plan

- Covered by existing test cases
- New test cases added

### CODE REVIEW

File: lib/private/glob_match.bzl [hunk 1]
```diff
@@ -13,31 +13,34 @@ def _split_on(expr, splits):
     # Splits an expression on the tokens in splits but keeps the tokens split in the result.
     # Tokens are matched in order so a token such as `**` should come before `*`.
     result = []
-    accumulator = ""
```
Comment (jbedard) [inline lib/private/glob_match.bzl:16]: There was a bug somewhere in here when doing the `**` work. This diff fixed the issue I was having and seems nicer imo 🤷

**Review by thesayyn — APPROVED** [review]
LGTM

File: lib/tests/glob_match_test.bzl [hunk 2]
```diff
@@ -166,6 +200,26 @@ def _mixed_wrapping_globstar(ctx):
 
 mixed_wrapper_globstar_test = unittest.make(_mixed_wrapping_globstar)
 
+def _all_of_ext(ctx):
+    return _glob_match_test(
+        ctx,
+        "**/*.tf",
+        matches = ["a/b.tf", "ab/cd/e.tf"],
```
Comment (gregmagolan) [inline lib/tests/glob_match_test.bzl:207]: This was the one that broke the release right? It should also match a .tf file at the root: `"c.tf"`

Comment (jbedard) [reply lib/tests/glob_match_test.bzl:207]: Yeah, and this PR doesn't fix it. This just does some prefactors before the fix.

**Pull request merged**